src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -126,6 +130,15 @@ void AbstractLQPNode::set_child(LQPChildSide side, const std::shared_ptr<Abstrac
   _child_changed();
 }
 
+size_t AbstractLQPNode::child_count() const {
+  /**
+   * Testing the shared_ptrs for null in _children to determine child count
+   */
+  return _children[0] ? (_children[1] ? 2 : 1) : (_children[1] ? 1 : 0);
+}
+
+bool AbstractLQPNode::is_leaf() const { return child_count() == 0; }
```

Comment:
since we decided the LQP is a graph and not a tree, we shouldn't talk about "leafs". Consider removing it (you're just using it in one Assert) - or I'll rename it once the children of the LQP become "inputs" and the parents "output"